File: app/src/main/java/com/parabola/newtone/mvp/presenter/fx/TabEqualizerPresenter.java
package com.parabola.newtone.mvp.presenter.fx;

import com.parabola.domain.interactor.player.AudioEffectsInteractor;
import com.parabola.newtone.di.app.AppComponent;
import com.parabola.newtone.mvp.view.fx.TabEqualizerView;
import com.parabola.newtone.ui.router.MainRouter;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import moxy.InjectViewState;
import moxy.MvpPresenter;

@InjectViewState
public final class TabEqualizerPresenter extends MvpPresenter<TabEqualizerView> {

    @Inject AudioEffectsInteractor fxInteractor;
    @Inject MainRouter router;

    private final CompositeDisposable disposables = new CompositeDisposable();

    public TabEqualizerPresenter(AppComponent appComponent) {
        appComponent.inject(this);
    }

    @Override
    public void onFirstViewAttach() {
        getViewState().setMaxEqLevel(fxInteractor.getMaxEqBandLevel());
        getViewState().setMinEqLevel(fxInteractor.getMinEqBandLevel());
        getViewState().refreshBands(fxInteractor.getBands());
        disposables.addAll(
                observeEqEnabling());
    }

    @Override
    public void onDestroy() {
        disposables.dispose();
    }

    private Disposable observeEqEnabling() {
        return fxInteractor.observeEqEnabling()
                .subscribe(enabled -> {
                    getViewState().setEqChecked(enabled);
                    getViewState().refreshBands(fxInteractor.getBands());
                });
    }


    public void onClickEqSwitcher(boolean enabled) {
        fxInteractor.setEqEnable(enabled);
    }

    public void onChangeBandLevel(int bandId, int newLevel) {
        fxInteractor.setBandLevel(bandId, newLevel);
    }

    public void onClickShowPresets() {
        router.openEqPresetsSelectorDialog();
    }

}

File: app/src/main/java/com/parabola/newtone/mvp/presenter/fx/TabEqualizerPresenter.kt
package com.parabola.newtone.mvp.presenter.fx

import com.parabola.domain.interactor.player.AudioEffectsInteractor
import com.parabola.newtone.di.app.AppComponent
import com.parabola.newtone.mvp.view.fx.TabEqualizerView
import com.parabola.newtone.ui.router.MainRouter
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import moxy.InjectViewState
import moxy.MvpPresenter
import javax.inject.Inject

@InjectViewState
class TabEqualizerPresenter(appComponent: AppComponent) : MvpPresenter<TabEqualizerView>() {

    @Inject
    lateinit var fxInteractor: AudioEffectsInteractor

    @Inject
    lateinit var router: MainRouter

    private val disposables = CompositeDisposable()


    init {
        appComponent.inject(this)
    }


    public override fun onFirstViewAttach() {
        viewState.setMaxEqLevel(fxInteractor.maxEqBandLevel.toInt())
        viewState.setMinEqLevel(fxInteractor.minEqBandLevel.toInt())
        viewState.refreshBands(fxInteractor.bands)
        disposables.addAll(observeEqEnabling())
    }

    override fun onDestroy() {
        disposables.dispose()
    }


    private fun observeEqEnabling(): Disposable {
        return fxInteractor.observeEqEnabling()
            .subscribe { enabled ->
                viewState.setEqChecked(enabled)
                viewState.refreshBands(fxInteractor.bands)
            }
    }


    fun onClickEqSwitcher(enabled: Boolean) {
        fxInteractor.setEqEnable(enabled)
    }

    fun onChangeBandLevel(bandId: Int, newLevel: Short) {
        fxInteractor.setBandLevel(bandId, newLevel)
    }

    fun onClickShowPresets() {
        router.openEqPresetsSelectorDialog()
    }

}
